src/genai/streamlit_pages/eyfs_kb_bbc_page.py

```diff
@@ -0,0 +1,178 @@
+        "src/genai/eyfs/prompts/constraints.json",
+        "src/genai/eyfs/prompts/situation.json",
+    ]
+
+    messages = [read_json(path) for path in paths]
+
+    # Get the user input
+    query = st.text_input(
+        label="**What's the topic you want activities for?**",
+        value="Let's create activities educating children on how whales breath",
+        help="Prompt the large language model with a some text and it will generate an activity plan for you.",
+    )
+
+    if st.button(label="**Generate**", help="Generate an answer."):
+        with st.spinner("Searching for relevant BBC activities..."):
+            # Encode the query
+            encoded_query = get_embedding(query)
+
+            # Search with Chroma
+            similar_docs = query_pinecone(
+                index,
+                encoded_query,
+                areas_of_learning=areas_of_learning,
+                top_n=4,
+                max_n=4,
+            )
+
+        with st.spinner("Generating activities..."):
+            # Create the prompt
+            messages_placeholders = {
+                "description": query,
+                "areas_of_learning": areas_of_learning,
+                "n_results": n_results,
+                "location": location,
+                "areas_of_learning_text": areas_of_learning_text,
+                "activity_examples": "\n======\n".join(
+                    [similar_doc["metadata"]["text"] for similar_doc in similar_docs]
+                ),
+            }
+
+            r = ActivityGenerator.generate(
+                model=selected_model,
+                temperature=temperature,
+                messages=messages,
+                message_kwargs=messages_placeholders,
+            )
+
+            st.write(r)
+
+        st.subheader("Sources")
+
+        for similar_doc in similar_docs:
+            title = similar_doc["metadata"]["title"]
+            url = similar_doc["id"]
+            category = similar_doc["metadata"]["areas_of_learning"]
+            st.write(f"""- [{title}]({url}) {category}""")
+
+
+@st.cache_resource
+def get_index(index_name: str, environment: str = "us-west1-gcp") -> pinecone.index.Index:
+    """Return and persist the pinecone index."""
+    pinecone.init(api_key=os.environ["PINECONE_API_KEY"], environment=environment)
+    index = pinecone.Index(index_name)
+    return index
+
+
+def try_sample_docs(num_docs: int, n: int) -> Union[list, ValueError]:
+    """Sample docs (without replacement)."""
+    try:
+        return random.sample(range(num_docs), n)
+    except ValueError as e:
+        return e
+
+
+def sample_docs(num_docs: int, n: int) -> Union[list, ValueError]:
+    """Sample docs (without replacement)."""
+    idx = try_sample_docs(num_docs, n)
+    if isinstance(idx, ValueError):
+        idx = sample_docs(num_docs, num_docs)
+        if isinstance(idx, ValueError):
+            raise ValueError(f"Cannot sample docs: {idx}")
+    return idx
```

Comment:
Interesting design, that you're trying first `n` samples and then sampling the total number `num_docs` - what's the rationale of this design versus simply sampling the `num_docs` and catching errors if there are any?

In any case, perhaps makes sense to define a default value of `n` ?